File: python/sorting/radix_sort.py
def counting_sort_for_radix(arr, exp):
    n = len(arr)
    
    # Output array
    output = [0] * n
    
    # Count array for digits (0–9)
    count = [0] * 10

    # Count occurrences of digits
    for num in arr:
        index = (num // exp) % 10
        count[index] += 1

    # Convert count[i] -> position in output
    for i in range(1, 10):
        count[i] += count[i - 1]

    # Build output array (stable sort: traverse from right)
    for i in range(n - 1, -1, -1):
        index = (arr[i] // exp) % 10
        output[count[index] - 1] = arr[i]
        count[index] -= 1

    # Copy back to original array
    for i in range(n):
        arr[i] = output[i]


def radix_sort(arr):
    if not arr:
        return arr

    # Find maximum to know number of digits
    max_val = max(arr)

    # Apply counting sort for each digit place
    exp = 1
    while max_val // exp > 0:
        counting_sort_for_radix(arr, exp)
        exp *= 10


# Example
arr = [170, 45, 75, 90, 802, 24, 2, 66]
print("Original:", arr)
radix_sort(arr)
print("Sorted:", arr)

File: cpp/sorting/radix_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

void countingSortForRadix(vector<int> &arr, int exp){
    int n = arr.size();
    vector<int> output(n);
    int count[10] = {0};

    for(int i=0; i<n; i++)
        count[(arr[i]/exp)%10]++;

    for(int i=1; i<10; i++)
        count[i] += count[i-1];

    for(int i=n-1; i>=0; i--){
        int index = (arr[i]/exp)%10;
        output[count[index]-1] = arr[i];
        count[index]--;
    }

    for(int i=0; i<n; i++)
        arr[i] = output[i];
}

void radixSort(vector<int> &arr){
    if(arr.empty()) return;
    int max_val = *max_element(arr.begin(), arr.end());
    for(int exp=1; max_val/exp>0; exp*=10)
        countingSortForRadix(arr, exp);
}

void printArray(const vector<int> &arr){
    for(int num: arr) cout << num << " ";
    cout << endl;
}

int main(){
    vector<int> arr = {170, 45, 75, 90, 802, 24, 2, 66};
    cout << "Original: ";
    printArray(arr);

    radixSort(arr);
    cout << "Radix Sorted: ";
    printArray(arr);

    return 0;
}

